### PR TITLE
[Merged by Bors] - Fix color banding by dithering image before quantization

### LINKED ISSUE #5262: Color Banding
## Bevy version

main

## What you did

Run the minimal `3d_scene` example.

## What went wrong

Very visible color banding:
image

## Additional information

This is a known limitation of rendering to 8 bit color channels. A common solution appears to be a debanding shader. [See here](https://docs.godotengine.org/en/stable/tutorials/3d/3d_rendering_limitations.html) for a related section in the Godot docs. While this isn't a bug in the sense that the output is *wrong*, it's a bug in that it *is* a well known visible artifact that we can mitigate.

### PR DESCRIPTION
**Pull request by aevyrie**
# Objective

- Closes #5262 
- Fix color banding caused by quantization.

## Solution

- Adds dithering to the tonemapping node from #3425.
- This is inspired by Godot's default "debanding" shader: https://gist.github.com/belzecue/c4b0bdede4e6d4b948679011096be150
- Unlike Godot:
  - debanding happens after tonemapping. My understanding is that this is preferred, because we are running the debanding at the last moment before quantization (`[f32, f32, f32, f32]` -> `f32`). This ensures we aren't biasing the dithering strength by applying it in a different (linear) color space.
  - This code instead uses and reference the origin source, Valve at GDC 2015

Screenshot from 2022-11-10 13-44-46
Screenshot from 2022-11-10 13-41-11

## Additional Notes 

Real time rendering to standard dynamic range outputs is limited to 8 bits of depth per color channel. Internally we keep everything in full 32-bit precision (`vec4<f32>`) inside passes and 16-bit between passes until the image is ready to be displayed, at which point the GPU implicitly converts our `vec4<f32>` into a single 32bit value per pixel, with each channel (rgba) getting 8 of those 32 bits.

### The Problem

8 bits of color depth is simply not enough precision to make each step invisible - we only have 256 values per channel! Human vision can perceive steps in luma to about 14 bits of precision. When drawing a very slight gradient, the transition between steps become visible because with a gradient, neighboring pixels will all jump to the next "step" of precision at the same time.

### The Solution

One solution is to simply output in HDR - more bits of color data means the transition between bands will become smaller. However, not everyone has hardware that supports 10+ bit color depth. Additionally, 10 bit color doesn't even fully solve the issue, banding will result in coherent bands on shallow gradients, but the steps will be harder to perceive.

The solution in this PR adds noise to the signal before it is "quantized" or resampled from 32 to 8 bits. Done naively, it's easy to add unneeded noise to the image. To ensure dithering is correct and absolutely minimal, noise is adding *within* one step of the output color depth. When converting from the 32bit to 8bit signal, the value is rounded to the nearest 8 bit value (0 - 255). Banding occurs around the transition from one value to the next, let's say from 50-51. Dithering will never add more than +/-0.5 bits of noise, so the pixels near this transition might round to 50 instead of 51 but will never round more than one step. This means that the output image won't have excess variance:
  - in a gradient from 49 to 51, there will be a step between each band at 49, 50, and 51.
  - Done correctly, the modified image of this gradient will never have a adjacent pixels more than one step (0-255) from each other.
  - I.e. when scanning across the gradient you should expect to see:
```
                  |-band-| |-band-| |-band-|
Baseline:         49 49 49 50 50 50 51 51 51
Dithered:         49 50 49 50 50 51 50 51 51
Dithered (wrong): 49 50 51 49 50 51 49 51 50
```

Screenshot from 2022-11-10 14-12-36
Screenshot from 2022-11-10 14-11-48

You can see from above how correct dithering "fuzzes" the transition between bands to reduce distinct steps in color, without adding excess noise.

### HDR

The previous section (and this PR) assumes the final output is to an 8-bit texture, however this is not always the case. When Bevy adds HDR support, the dithering code will need to take the per-channel depth into account instead of assuming it to be 0-255. Edit: I talked with Rob about this and it seems like the current solution is okay. We may need to revisit once we have actual HDR final image output.

---

### Added

- All pipelines now support deband dithering. This is enabled by default in 3D, and can be toggled in the `Tonemapping` component in camera bundles. Banding is a graphical artifact created when the rendered image is crunched from high precision (f32 per color channel) down to the final output (u8 per channel in SDR). This results in subtle gradients becoming blocky due to the reduced color precision. Deband dithering applies a small amount of noise to the signal before it is "crunched", which breaks up the hard edges of blocks (bands) of color. Note that this does not add excess noise to the image, as the amount of noise is less than a single step of a color channel - just enough to break up the transition between color blocks in a gradient.

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
Please note that this fixes #5262 :)

**Comment by aevyrie** [comment]
FYI I'm planning on making an example to better stress test noise and banding performance before removing draft status.

File: crates/bevy_pbr/src/render/pbr.wgsl [hunk 1]
```diff
@@ -98,8 +106,10 @@ fn fragment(in: FragmentInput) -> [[location(0)]] vec4<f32> {
...
 
-        output_color = tone_mapping(pbr(pbr_input));
+        let pbr_out = pbr(pbr_input);
+        output_color = screen_space_dither(in.frag_coord.xy) * min(15.0 * pbr_out, vec4<f32>(1.0)) + pbr_out;
+        output_color = tone_mapping(output_color);
     }
 
     return output_color;
-}
+}
```
Comment (bjorn3) [inline crates/bevy_pbr/src/render/pbr.wgsl:115]: nit: missing trailing newline

File: crates/bevy_pbr/src/render/pbr.wgsl [hunk 2]
```diff
@@ -24,6 +24,14 @@ struct FragmentInput {
 #endif
 };
 
+fn screen_space_dither(frag_coord: vec2<f32>) -> vec4<f32> {
+	// Iestyn's RGB dither (7 asm instructions) from Portal 2 X360, slightly modified for VR.
+    var dither = vec3<f32>(dot(vec2<f32>(171.0, 231.0), frag_coord));
+    dither = fract(dither.rgb / vec3<f32>(103.0, 71.0, 97.0));
+	// Subtract 0.5 to avoid slightly brightening the whole viewport.
+    return vec4<f32>((dither.rgb - 0.5) / 255.0, 0.0);
+}
```
Comment (bjorn3) [inline crates/bevy_pbr/src/render/pbr.wgsl:33]: Where did this code originate from? Did you modify it for VR yourself, or did someone else do so?

Comment (aevyrie) [reply crates/bevy_pbr/src/render/pbr.wgsl:33]: Another reason this is marked as draft - I haven't looked at the licenses for any of my references. Might need some support for that. Unsure if little code snippets like this are acceptable.

File: crates/bevy_pbr/src/render/pbr.wgsl [hunk 3]
```diff
@@ -98,8 +106,10 @@ fn fragment(in: FragmentInput) -> [[location(0)]] vec4<f32> {
         );
         pbr_input.V = calculate_view(in.world_position, pbr_input.is_orthographic);
 
-        output_color = tone_mapping(pbr(pbr_input));
+        let pbr_out = pbr(pbr_input);
+        output_color = screen_space_dither(in.frag_coord.xy) * min(15.0 * pbr_out, vec4<f32>(1.0)) + pbr_out;
+        output_color = tone_mapping(output_color);
```
[the remark this answers is not in the document]
Comment (aevyrie) [reply crates/bevy_pbr/src/render/pbr.wgsl]: Thanks for the heads up. I'll probably just rebase on that and add a debanding node, or put it in tonemap.

**Comment by aevyrie** [comment]
Added a stress test for banding:
### Control
image
### Dithered
image

**Comment by aevyrie** [comment]
@CptPotato I rebased on #3425, and added it to the tonemapping pass. This means dithering is now a fullscreen post-process. Could probably improve by moving to its own node that outputs to a `Bgra8Unorm` (instead of SRGB, I think this is correct?) to remove the unnecessary conversion to and from SRGB space.

**Comment by devil-ira** [comment]
As a note to others viewing the images above if you still see banding in the dithered image click on it to view it in another tab.
That will display the image at 100% scaling and the banding should not be visible.

**Comment by aevyrie** [comment]
> The result looks good! If tonemapping is disabled, will dithering be disabled too?

Yup, that's correct. Maybe that make sense though? The goal of dithering is to remove artifacts from quantization, which seems to fit alongside tonemapping (mapping between color spaces).

> [...] you could probably eliminate the conversion back to linear by changing the format of the render target view ("attachment" I think is the wgpu slang).

The issue with this is device compatibility from what @superdump mentioned on discord.

> Though, in that case a more accurate conversion function than `pow(col, 1.0/2.2)` would make sense.

Any resources on what that would be?

**Comment by aevyrie** [comment]
Thanks for the links!

> That being said, if you keep the conversion in both directions (linear -> srgb -> linear) in the shader the difference it makes on the dithering may be negligible. I can't say for sure, though.

This is a good point. Dithering just relies on the rounding being correct, such that dithering never causes more than one full step away from the true signal value during quantization. As long as the log approximation is "pretty close" (a rigorous technical term), it wouldn't cause quantization to round another full step for the large majority of cases.

**Comment by superdump** [comment]
> > [...] you could probably eliminate the conversion back to linear by changing the format of the render target view ("attachment" I think is the wgpu slang).
> 
> The issue with this is device compatibility from what @superdump mentioned on discord.

This is only relevant if rendering directly into the swapchain texture. The formats supported by the swapchain are more limited as they're specifically for outputting to a display. If rendering to an intermediate texture then the format support is much broader. The tonemapping / upscaling PR does allow for using an independent intermediate texture rather than rendering directly to the swapchain texture, so in some cases at least we could do this. Basically probably only the final swapchain texture might need to be an srgb format and we can use non-srgb in-between. We could even do gamma correction manually and not use an srgb format at all, though I'm not sure of exact swapchain texture format support across a variety of different backends / GPU vendors though so it may be that some/many only support srgb formats.

**Comment by superdump** [comment]
I asked Alex Vlachos about the license of the GDC 2015 code: https://twitter.com/AlexVlachos/status/1557662473775239175

**Comment by aevyrie** [comment]
@mockersf and @cart, RE: discord, I've cleaned this up and got it working with and without hdr.

**Comment by superdump** [comment]
I've reloaded this information into my mental cache and had a think through the current state of things.

SDR:
* main pass f32 with tonemapping -> srgb u8 texture
* Dithering should happen in the main pass on non-linear sRGB values, then be converted back to linear just before writing out

'HDR' (it's still SDR at the end as we don't have HDR output to the window yet, but it allows us to do post-processing on linear HDR before tonemapping):
* main pass f32 -> linear f16 texture -> tonemapping f32 -> linear f16 texture -> upscaling f32 -> srgb u8 texture
* Dithering should happen in the upscaling pass on non-linear sRGB values, then be converted back to linear just before writing out

I think this should be correct...

**Comment by superdump** [comment]
Nope, I forgot about upscaling in the SDR case:
* main pass f32 with tonemapping -> srgb u8 texture -> upscaling f32 -> srgb u8 texture
* Does dithering need to be done twice to compensate for both the intermediate quantisation errors and the upscaled errors? i.e. in both the main pass and upscaling pass?

**Comment by cart** [comment]
This should be disable-able (probably on the existing Tonemapping settings, given that they are currently tied together). Disabling dither there should specialize the tonemapping pipeline to remove the dither code.

**Comment by aevyrie** [comment]
I'm not super proficient with pipeline specialization, but I think I've completed this as requested. Dithering can be toggled, but requires tonemapping to be enabled. This is reflected in the new type signature of `Tonemapping`.

File: crates/bevy_core_pipeline/src/tonemapping/tonemapping_shared.wgsl [hunk 4]
```diff
@@ -27,3 +27,11 @@ fn reinhard_luminance(color: vec3<f32>) -> vec3<f32> {
...
     return tonemapping_change_luminance(color, l_new);
 }
+
+// Source: Advanced VR Rendering, GDC 2015, Alex Vlachos, Valve, Slide 49
+// https://media.steampowered.com/apps/valve/2015/Alex_Vlachos_Advanced_VR_Rendering_GDC2015.pdf
+fn screen_space_dither(frag_coord: vec2<f32>) -> vec3<f32> {
+    var dither = vec3<f32>(dot(vec2<f32>(171.0, 231.0), frag_coord)).xxx;
+    dither = fract(dither.rgb / vec3<f32>(103.0, 71.0, 97.0));
+    return (dither - 0.5) / 255.0;
+}
```
Comment (mockersf) [inline crates/bevy_core_pipeline/src/tonemapping/tonemapping_shared.wgsl:37]:
```suggestion
}

```

**Comment by mockersf** [comment]
just checking, as it's not what I understood from the discussion: should this dithering setting be ignored when HDR is enabled?

**Comment by cart** [comment]
> just checking, as it's not what I understood from the discussion: should this dithering setting be ignored when HDR is enabled?

Nope. Dithering should be coupled to tonemapping (at least for now, as they are logically coupled in the hdr tonemapping pass). If HDR is enabled, dithering should happen in the tonemapping pass. If HDR is disabled, it should happen in the PBR / sprite shader.

This impl doesn't quite line up with that, as we aren't specializing the tonemapping pipeline when dither is enabled, which means it is skipped when hdr is enabled.

**Comment by cart** [comment]
Working on a fix now

**Comment by cart** [comment]
Fixed! I also made a personal style call by renaming `is_deband_dither_enabled` to `deband_dither`. I like preferring the shorter names when the word is already a verb. I think they read better by nature of having less grammar to mentally parse.

**Comment by aevyrie** [comment]
> Fixed! I also made a personal style call by renaming `is_deband_dither_enabled` to `deband_dither`. I like preferring the shorter names for bools when the word is already a verb. I think they read better by nature of having less grammar to mentally parse.

Thanks. I agree with the rename, I had the same initially but chose to follow the preexisting `is_` style. :slightly_smiling_face:

File: crates/bevy_core_pipeline/src/tonemapping/tonemapping.wgsl [hunk 5]
```diff
@@ -10,5 +10,15 @@ var hdr_sampler: sampler;
 fn fragment(in: FullscreenVertexOutput) -> @location(0) vec4<f32> {
     let hdr_color = textureSample(hdr_texture, hdr_sampler, in.uv);
 
-    return vec4<f32>(reinhard_luminance(hdr_color.rgb), hdr_color.a);
+    var output_rgb = vec3<f32>(reinhard_luminance(hdr_color.rgb));
```
Comment (superdump) [inline crates/bevy_core_pipeline/src/tonemapping/tonemapping.wgsl:13]: Nit: if Reinhard luminance returns a vec3 then wrapping it in one is unnecessary.

**Review by superdump — APPROVED** [review]
I’m marking this as approved assuming it is all tested in its current state as I only had a couple of nit comments.

File: crates/bevy_core_pipeline/src/tonemapping/tonemapping.wgsl (hunk 5 above)
Comment (aevyrie) [reply crates/bevy_core_pipeline/src/tonemapping/tonemapping.wgsl:13]: The intermediate variable is needed to work with the ifdef, and reassemble a vec4 with the alpha value. Right?

Comment (superdump) [reply crates/bevy_core_pipeline/src/tonemapping/tonemapping.wgsl:13]: I wasn’t suggesting to remove the variable. I was just saying that if the function returns a vec3 then we don’t need to do vec3(func()), just func().

Comment (aevyrie) [reply crates/bevy_core_pipeline/src/tonemapping/tonemapping.wgsl:13]: Oh, duh! Thanks for clarifying. 🙂

**Pull request closed without merging**